docs/source/references/schema.json

```
@@ -631,7 +745,25 @@
       }
     ]
   },
-  "ConfigurationEventSet": "u32",
+  "ConfigurationEventSet": {
+    "Bitmap": {
+      "repr": "u32",
+      "masks": [
+        {
+          "name": "Changed",
+          "mask": 1
+        },
+        {
+          "name": "Created",
+          "mask": 2
+        },
+        {
+          "name": "Deleted",
+          "mask": 4
+        }
+      ]
+    }
+  },
```

```suggestion
  "ConfigurationEventSet": {
    "Enum": [
      {
        "tag": "Changed",
        "discriminant": 1
      },
      {
        "tag": "Created",
        "discriminant": 2
      },
      {
        "tag": "Deleted",
        "discriminant": 4
      }
    ]
  },
```

it seems to me that we could have just treated it as an enum. I don't like seeing extra complexity in the schema

---

@0x009922 how will you use this in the sdk? does current schema format bring you something that enum doesn't?

---

I think re-using enum in this case would be more problematic than helpful. After all, enums variants are mutually exclusive, while this new structure could have many values set at once.

If we use a enum in this case, it will enforce a hard code on SDK codegen side, saying "this specific enum is not actually a enum". I would rather have a separate structure for bit masks explicitly in the schema.

---

> If we use a enum in this case, it will enforce a hard code on SDK codegen side, saying "this specific enum is not actually a enum". I would rather have a separate structure for bit masks explicitly in the schema.

hm, I respectfully disagree here. This is not something like an enum, this is literally an enum, at least from the perspective of a schema. It's an implementation detail how this enum is actually represented in the code. 

In Rust we cannot express relations (`BitOr` and `BitOrAssign`) between variants of this enum as we think is the most ergonomical, so we fallback to using a combination of a struct(that is made to look like an enum) and associated constants to represent this type. However, in a different language (think of `Haskell` where enum variants are types in their own right) it might be possible to represent this type as an enum

---

> If we use a enum in this case, it will enforce a hard code on SDK codegen side, saying "this specific enum is not actually a enum". I would rather have a separate structure for bit masks explicitly in the schema.

hm, I respectfully disagree here. This is not something like an enum, this is literally an enum, at least from the perspective of a schema. It's an implementation detail how this enum is actually represented in the code.

@DCNick3 correct me if I'm wrong, but you could have implemented this type as a regular enum, but you made an optimization. Whether or not an SDK will make this optimization is not something to be required by a schema

---

`EventSet` is an optimization indeed, but it is not equivalent to an enum, but to a `HashSet<Enum>`. Enums have one and only one variant set at the same time, while `EventSet` can have any combination of set and unset variants. As we know that the set of `Enum` values is low we store them as single bits instead of allocating a whole `HashSet` on the heap.

What `EventSet`s currently have though, is a corresponding enum that can be used to reconstruct the set of masks using the same logic as the derive macro. However, the logic is non-trivial (there are some name mappings applied to the mappings, for example) and would have to duplicate the logic of the derive macro. Exposing it in schema keeps the translation logic in one place (derive macro) and makes SDK codegen straightforward, so I think schema is _the_ place to have this information.

---

ok, I concede. This is the best solution :)